Before you turn this homework in, make sure everything runs as expected. First, **restart the kernel** and then **run all cells** (in the menubar, select Kernel$\rightarrow$Restart \& Run all). If you get error messages, the grader will get the same errors and won't be able to grade your work. 

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE". You may delete the line(s) that say "**raise** NotImplementedError()" and replace them with your actual work. Write all your work in the cells provided in this notebook. **DO NOT** add or delete any cells. 

Please fill in your name in the cell below.

In [1]:
NAME = "Zander Bonnet"

---

## Problem 1

The file "Temp.txt" which you can download from the homework tab on Canvas, contains the monthly high-temperatures (in degrees Celsius) at Heathrow airport for the years 1948 through 2016. Save the file on your computer (in a folder that you can find again). Run the code below, changing the path to the appropriate path on **your** computer. It's ok to use forward slashes in the path name, regardless of whether you're working on Windows or MacOS. 

In [2]:
temperatures = []

from pathlib import Path

data_folder = Path('/Users/zanderbonnet/Desktop/Math 167PS/Homework/Data')
# make sure to change this path to the actual folder where YOU have saved the data file. 

file_to_open = data_folder / "Temp.txt"

file_to_open
f = open(file_to_open)

for row in f:
    temperatures.append(float(row))

You have now created an object called temperatures. 

Answer the following questions by inserting your code into the code box (after the "# your code here" prompt. Comment out or delete the "raise NotImplementedError()") and answering the question in the form of a complete English sentence in the respective Markdown box. 

a) What kind of object type is "temperatures"?

In [3]:
# YOUR CODE HERE
type(temperatures)

list

The object temperatures is a LIST.

b) Describe what the "float()" and "append()" commands do, respectively, that have been applied above. Try to leave out the "float()" command. What happens then? 

float() - This command takes the data input from every row of the .txt file and converts the it to a float. It is done becasue when we read off the file the data is originaly stored as a string.

append() - This command takes the the input of a data point read off of the file, after it has been converted to a float, and sticks it on the end of the list temperatures.

c) How many temperatures have been recorded in the file?

In [4]:
# YOUR CODE HERE
len(temperatures)

828

There are 828 temperatures recorded in the file.

d) What are the highest and lowest temperatures recorded?

In [5]:
# YOUR CODE HERE
print(min(temperatures))
print(max(temperatures))

0.8
28.2


The highest temperature recorded was 28.2 and the lowest was 0.8.

e) The first observation in the `temperatures` object is from a January 1948. Using only the commands you learned from chapter 5 (that is, using base python and no additional python libraries), find the the mean January high temperature at London Heathrow between the years 1948 and 2016. Round your answer to two decimal digits accuracy. 

In [6]:
# YOUR CODE HERE
janTemps = temperatures[::12]
mean = sum(janTemps)/len(janTemps)
print(round(mean,2))

7.56


The average temperature in January from 1948 to 2016 was 7.56.

f) Find the median monthly high temperature (using only the commands you learned from chapter 5 and base python, not using additional modules or libraries). Use Python to round your result to two after-decimal digits. 

In [7]:
# YOUR CODE HERE
sortedTemps = sorted(temperatures)
#sortedTemps = [1,2,5,7,8,9]
mid = len(sortedTemps)/2
if (mid%1) == 0:
    upper = sortedTemps[int(mid)]
    lower = sortedTemps[int(mid - 1)]
    print(round((upper + lower)/2,2))
else:
    print(round(sortedTemps[int(mid)],2))

14.7


The median monthly high temperature was 14.7.

## Problem 2 

Much of data science these days is based on text data. The text to be analyzed may be literature or postings on social media. The tasks to be performed could be sentiment analysis (trying to guess whether the writing is "pro" or "con" on a certain issue) to automatically detecting postings by Malicious bots. However, before you can perform any kind of machine learning procedure on a text, you have to first learn how to "clean" the text. That usually means removing all non-text elements and sometimes also removing filler words such as "the", "it", "and" etc. After that, what is usually analyzed is the frequency with which certain words appear and most often actually not the *order* in which they appear. 

The text below are the first three paragraphs of Franz Kafka's book "Metamorphosis". Our goal is to clean this text passage. 

In [8]:
text = """One morning, when Gregor Samsa woke from troubled dreams, he found
himself transformed in his bed into a horrible vermin.  He lay on
his armour-like back, and if he lifted his head a little he could
see his brown belly, slightly domed and divided by arches into stiff
sections.  The bedding was hardly able to cover it and seemed ready
to slide off any moment.  His many legs, pitifully thin compared
with the size of the rest of him, waved about helplessly as he
looked.

"What's happened to me?" he thought.  It wasn't a dream.  His room,
a proper human room although a little too small, lay peacefully
between its four familiar walls.  A collection of textile samples
lay spread out on the table - Samsa was a travelling salesman - and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice, gilded frame.  It showed
a lady fitted out with a fur hat and fur boa who sat upright,
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer.

Gregor then turned to look out the window at the dull weather.
Drops of rain could be heard hitting the pane, which made him feel
quite sad.  "How about if I sleep a little bit longer and forget all
this nonsense", he thought, but that was something he was unable to
do because he was used to sleeping on his right, and in his present
state couldn't get into that position.  However hard he threw
himself onto his right, he always rolled back to where he was.  He
must have tried it a hundred times, shut his eyes so that he
wouldn't have to look at the floundering legs, and only stopped when
he began to feel a mild, dull pain there that he had never felt
before."""

a) Look at the text. What non-alphabetic character elements do you see in the text?

The non-alphabetic characters are [" , "  , " - " , " . " , " ? " , " " "  , " ' "]

b) Using only the commands covered in class so far (in base python without use of additional libraries), write code to remove all non-alphabetic characters (such as punctuation).

In [9]:
# YOUR CODE HERE
nonAlph = [",","-",".","'","?",'"']
for x in nonAlph:
    text = text.replace(x,"")
print(text)

One morning when Gregor Samsa woke from troubled dreams he found
himself transformed in his bed into a horrible vermin  He lay on
his armourlike back and if he lifted his head a little he could
see his brown belly slightly domed and divided by arches into stiff
sections  The bedding was hardly able to cover it and seemed ready
to slide off any moment  His many legs pitifully thin compared
with the size of the rest of him waved about helplessly as he
looked

Whats happened to me he thought  It wasnt a dream  His room
a proper human room although a little too small lay peacefully
between its four familiar walls  A collection of textile samples
lay spread out on the table  Samsa was a travelling salesman  and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice gilded frame  It showed
a lady fitted out with a fur hat and fur boa who sat upright
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer

Gregor 

c) Write code to make everything lower case and convert the text into a list of single words. 

In [10]:
# YOUR CODE HERE
wordList = text.split()
for i in range(len(wordList)):
    wordList[i] = wordList[i].lower()  
print(wordList)

['one', 'morning', 'when', 'gregor', 'samsa', 'woke', 'from', 'troubled', 'dreams', 'he', 'found', 'himself', 'transformed', 'in', 'his', 'bed', 'into', 'a', 'horrible', 'vermin', 'he', 'lay', 'on', 'his', 'armourlike', 'back', 'and', 'if', 'he', 'lifted', 'his', 'head', 'a', 'little', 'he', 'could', 'see', 'his', 'brown', 'belly', 'slightly', 'domed', 'and', 'divided', 'by', 'arches', 'into', 'stiff', 'sections', 'the', 'bedding', 'was', 'hardly', 'able', 'to', 'cover', 'it', 'and', 'seemed', 'ready', 'to', 'slide', 'off', 'any', 'moment', 'his', 'many', 'legs', 'pitifully', 'thin', 'compared', 'with', 'the', 'size', 'of', 'the', 'rest', 'of', 'him', 'waved', 'about', 'helplessly', 'as', 'he', 'looked', 'whats', 'happened', 'to', 'me', 'he', 'thought', 'it', 'wasnt', 'a', 'dream', 'his', 'room', 'a', 'proper', 'human', 'room', 'although', 'a', 'little', 'too', 'small', 'lay', 'peacefully', 'between', 'its', 'four', 'familiar', 'walls', 'a', 'collection', 'of', 'textile', 'samples', 'l

d) How many words are in the list? Count, how often the word "he" appears in the three paragraphs. Don't count if "he" is part of another word, e.g., "tHEn". 

In [11]:
# YOUR CODE HERE
print(len(wordList))
print(wordList.count("he"))

313
17


There are 313 words in the list and the word "he" appears 17 times.

e) How many distinct (different) words appear in the first three paragragraphs?

In [12]:
# YOUR CODE HERE
uniqueWords = set(wordList)
print(len(uniqueWords))


#Method 2
unWor = list()
for i in range(len(wordList)):
    if(wordList[i] not in unWor):
        unWor.append(wordList[i])
print(len(unWor))

194
194


There are 194 different words in the first three paragraphs.